File: swift/Sources/AetherProtocol/Reputation/BehavioralAnomalyDetector.swift
// SPDX-License-Identifier: MIT

import Foundation

// MARK: - Options

/// Tuning knobs for all three anomaly detectors.
///
/// All fields have production-safe defaults; pass a customised instance in
/// tests to exercise edge-cases without waiting for real-time windows.
public struct AnomalyDetectorOptions {
    /// Length of the rolling volume window in milliseconds (default 30 s).
    public var volumeWindowMs: Int64         = 30_000
    /// Packets-per-window multiplier over the EWMA baseline that trips a spike (default 5×).
    public var volumeSpikeMultiplier: Double = 5.0
    /// Exponential smoothing factor applied when a volume window rolls (default α = 0.20).
    public var ewmaAlpha: Double             = 0.20
    /// Sliding window over which unique destinations are counted (default 60 s).
    public var scatterWindowMs: Int64        = 60_000
    /// Unique-destination count that trips scatter detection (default 50).
    public var scatterThreshold: Int         = 50
    /// Number of leading geohash characters compared for location verification (default 4).
    public var geohashPrefixLength: Int      = 4
    /// Minimum gap in ms between consecutive geohash-mismatch penalties for the same UHID (default 60 s).
    public var geohashRateLimitMs: Int64     = 60_000

    public init() {}
}

// MARK: - Actor

/// Thread-safe behavioral anomaly detector that drives ``NodeReputationService``
/// based on three orthogonal signal sources:
///
/// 1. **Volume spike** — EWMA-based packet-rate anomaly per source UHID.
/// 2. **Destination scatter** — fan-out to an unusually large set of peers.
/// 3. **Geohash mismatch** — claimed location doesn't match routing evidence.
///
/// All detection is reactive (no background timers); call the `observe*` methods
/// from your packet-processing pipeline.
public actor BehavioralAnomalyDetector {

    // MARK: - Dependencies

    /// The reputation service that receives penalties when anomalies fire.
    public let reputation: NodeReputationService

    // MARK: - Configuration

    private let opts: AnomalyDetectorOptions

    // MARK: - Volume state

    /// Per-source rolling-window + EWMA state.
    private var volumeState: [String: VolumeState] = [:]

    private struct VolumeState {
        var windowStart: Int64   = 0
        var windowCount: Int     = 0
        var ewmaBaseline: Double = 0
        var hasBaseline: Bool    = false
    }

    // MARK: - Scatter state

    /// Per-source list of (destination, timestamp) observations within the scatter window.
    private var scatterEntries: [String: [(dest: String, ts: Int64)]] = [:]

    // MARK: - Geohash state

    /// Timestamp of the last geohash-mismatch penalty issued per UHID (for rate-limiting).
    private var geoLastSignal: [String: Int64] = [:]

    // MARK: - Initialiser

    public init(reputation: NodeReputationService,
                opts: AnomalyDetectorOptions = .init()) {
        self.reputation = reputation
        self.opts = opts
    }

    // MARK: - Public API

    /// Record a packet from `sourceUhid` to `destinationUhid` at `timestampMs`.
    ///
    /// Internally runs both the volume-spike check and the destination-scatter check.
    public func observePacket(sourceUhid: String,
                               destinationUhid: String,
                               timestampMs: Int64) async {
        await checkVolume(sourceUhid: sourceUhid, timestampMs: timestampMs)
        await checkScatter(sourceUhid: sourceUhid,
                           destinationUhid: destinationUhid,
                           timestampMs: timestampMs)
    }

    /// Compare `claimedGeohash` against the geohash inferred from routing
    /// (`observedRoutingGeohash`).  When the leading `geohashPrefixLength`
    /// characters differ, a ``NodeReputationService/recordSignatureFailure(uhid:)``
    /// is fired — subject to `geohashRateLimitMs` per-UHID throttling.
    ///
    /// - Parameter timestampMs: Caller-supplied logical timestamp (ms since epoch).
    ///   Accepting this explicitly keeps the method deterministically testable.
    public func observeGeohashClaim(uhid: String,
                                     claimedGeohash: String,
                                     observedRoutingGeohash: String,
                                     timestampMs: Int64) async {
        let claimedPrefix  = String(claimedGeohash.prefix(opts.geohashPrefixLength))
        let observedPrefix = String(observedRoutingGeohash.prefix(opts.geohashPrefixLength))

        guard claimedPrefix != observedPrefix else { return }

        // Rate-limit: suppress if a penalty was already issued within the window.
        let shouldFire: Bool
        if let last = geoLastSignal[uhid] {
            shouldFire = (timestampMs - last) >= opts.geohashRateLimitMs
        } else {
            shouldFire = true
        }

        if shouldFire {
            geoLastSignal[uhid] = timestampMs
            await reputation.recordSignatureFailure(uhid: uhid)
        }
    }

    /// Observe an SPK signature verification failure for `uhid`.
    ///
    /// This is a direct passthrough to ``NodeReputationService/recordSignatureFailure(uhid:)``.
    public func observeSpkSigFailure(uhid: String) async {
        await reputation.recordSignatureFailure(uhid: uhid)
    }

    // MARK: - Internal: volume-spike detection

    /// Rolling-window EWMA volume-spike check.
    ///
    /// **Algorithm:**
    /// - While the current timestamp is still within the active window, just
    ///   increment the counter.
    /// - When the window expires, compare the completed window's packet count
    ///   against the EWMA baseline:
    ///   - First completed window seeds the baseline (no penalty yet).
    ///   - Subsequent windows update the EWMA with α and fire a flood penalty
    ///     if `windowCount > multiplier × ewmaBaseline` (and baseline > 0).
    ///   - Reset window start and count for the new period.
    private func checkVolume(sourceUhid: String, timestampMs: Int64) async {
        var state = volumeState[sourceUhid] ?? VolumeState()

        if state.windowStart == 0 {
            // First packet ever from this source — open the first window.
            state.windowStart = timestampMs
            state.windowCount = 1
            volumeState[sourceUhid] = state
            return
        }

        let elapsed = timestampMs - state.windowStart

        if elapsed >= opts.volumeWindowMs {
            // Window has expired — evaluate then roll.
            let completedCount = state.windowCount

            if !state.hasBaseline {
                // Seed baseline from first completed window; no penalty.
                state.ewmaBaseline = Double(completedCount)
                state.hasBaseline  = true
            } else {
                // Update EWMA.
                let alpha    = opts.ewmaAlpha
                let oldEwma  = state.ewmaBaseline
                state.ewmaBaseline = alpha * Double(completedCount) + (1.0 - alpha) * oldEwma

                // Fire if spike is above threshold.
                if state.ewmaBaseline > 0,
                   Double(completedCount) > opts.volumeSpikeMultiplier * oldEwma {
                    await reputation.recordRreqFloodAttempt(uhid: sourceUhid)
                }
            }

            // Roll the window.
            state.windowStart = timestampMs
            state.windowCount = 1
        } else {
            state.windowCount += 1
        }

        volumeState[sourceUhid] = state
    }

    // MARK: - Internal: destination-scatter detection

    /// Sliding-window unique-destination fan-out check.
    ///
    /// Appends the current observation, prunes entries outside `scatterWindowMs`,
    /// then fires a flood penalty when the unique-destination count exceeds
    /// `scatterThreshold`.
    private func checkScatter(sourceUhid: String,
                               destinationUhid: String,
                               timestampMs: Int64) async {
        var entries = scatterEntries[sourceUhid] ?? []

        // Append new observation.
        entries.append((dest: destinationUhid, ts: timestampMs))

        // Prune entries that have aged out of the window.
        entries = entries.filter { timestampMs - $0.ts <= opts.scatterWindowMs }

        scatterEntries[sourceUhid] = entries

        // Count unique destinations in the pruned window.
        let uniqueCount = Set(entries.map(\.dest)).count
        if uniqueCount > opts.scatterThreshold {
            await reputation.recordRreqFloodAttempt(uhid: sourceUhid)
        }
    }
}

File: kotlin/src/main/kotlin/aethermesh/security/BehavioralAnomalyDetector.kt
// SPDX-License-Identifier: MIT

package aethermesh.security

import java.util.concurrent.ConcurrentHashMap

/**
 * Detects behavioural anomalies from network peers and feeds signals into
 * [NodeReputationService].
 *
 * Three detection axes:
 *  1. **Volume spike** — per-source packet rate vs. EWMA baseline.
 *  2. **Destination scatter** — unique destinations within a rolling window.
 *  3. **Geohash mismatch** — claimed location vs. observed routing location.
 *
 * All per-UHID state is thread-safe via fine-grained per-key lock objects
 * stored in [ConcurrentHashMap]s.
 */
class BehavioralAnomalyDetector(
    private val reputation: NodeReputationService,
    private val opts: AnomalyDetectorOptions = AnomalyDetectorOptions()
) {

    // ── Volume tracking ──────────────────────────────────────────────────────

    private data class VolumeState(
        var windowStart: Long = 0L,
        var windowCount: Int = 0,
        var ewmaBaseline: Double = 0.0,
        var hasBaseline: Boolean = false
    )

    /** Lock objects — one per source UHID. */
    private val volumeLocks = ConcurrentHashMap<String, Any>()

    /** Mutable volume state — one entry per source UHID. */
    private val volumeData = ConcurrentHashMap<String, VolumeState>()

    // ── Scatter tracking ─────────────────────────────────────────────────────

    /** Lock objects — one per source UHID. */
    private val scatterLocks = ConcurrentHashMap<String, Any>()

    /**
     * List of (destinationUhid, timestampMs) pairs seen within
     * [AnomalyDetectorOptions.scatterWindowMs].
     */
    private val scatterData = ConcurrentHashMap<String, MutableList<Pair<String, Long>>>()

    // ── Geohash rate limiting ────────────────────────────────────────────────

    /** Lock objects for atomic check-then-act in geohash mismatch handling. */
    private val geoLocks = ConcurrentHashMap<String, Any>()

    /** Timestamp (ms) of the last geohash-mismatch signal fired for each UHID. */
    private val geoLastSignal = ConcurrentHashMap<String, Long>()

    // ── Public API ───────────────────────────────────────────────────────────

    /**
     * Record that a packet was seen from [sourceUhid] bound for [destinationUhid]
     * at [timestampMs].  Runs volume-spike and scatter checks.
     */
    fun observePacket(sourceUhid: String, destinationUhid: String, timestampMs: Long) {
        checkVolume(sourceUhid, timestampMs)
        checkScatter(sourceUhid, destinationUhid, timestampMs)
    }

    /**
     * Record that [uhid] claimed [claimedGeohash] in its packet header but was
     * routed from a location whose geohash prefix is [observedRoutingGeohash].
     *
     * If the first [AnomalyDetectorOptions.geohashPrefixLength] characters
     * differ **and** the per-UHID rate limit has expired, fires
     * [NodeReputationService.recordSignatureFailure].
     */
    fun observeGeohashClaim(
        uhid: String,
        claimedGeohash: String,
        observedRoutingGeohash: String,
        timestampMs: Long = System.currentTimeMillis()
    ) {
        val claimedPrefix  = claimedGeohash.take(opts.geohashPrefixLength)
        val observedPrefix = observedRoutingGeohash.take(opts.geohashPrefixLength)

        if (claimedPrefix == observedPrefix) return

        // Rate-limit: at most one signal per geohashRateLimitMs per UHID.
        val lock = geoLocks.computeIfAbsent(uhid) { Any() }
        synchronized(lock) {
            val lastSignal = geoLastSignal[uhid]
            if (lastSignal == null || timestampMs - lastSignal >= opts.geohashRateLimitMs) {
                geoLastSignal[uhid] = timestampMs
                reputation.recordSignatureFailure(uhid)
            }
        }
    }

    /**
     * Direct passthrough: record that an SPK (Ed25519) signature verification
     * failed for a packet from [uhid].
     */
    fun observeSpkSigFailure(uhid: String) {
        reputation.recordSignatureFailure(uhid)
    }

    // ── Private helpers ──────────────────────────────────────────────────────

    private fun checkVolume(sourceUhid: String, timestampMs: Long) {
        val lock  = volumeLocks.computeIfAbsent(sourceUhid) { Any() }
        val state = volumeData.computeIfAbsent(sourceUhid) { VolumeState(windowStart = timestampMs) }

        synchronized(lock) {
            if (timestampMs - state.windowStart >= opts.volumeWindowMs) {
                // ── Roll the window ──────────────────────────────────────────
                if (!state.hasBaseline) {
                    // First roll: seed EWMA from this window's count
                    state.ewmaBaseline = state.windowCount.toDouble()
                    state.hasBaseline  = true
                } else {
                    // Update EWMA
                    state.ewmaBaseline =
                        opts.ewmaAlpha * state.windowCount + (1.0 - opts.ewmaAlpha) * state.ewmaBaseline

                    // Spike check — only when EWMA is positive to avoid ÷0
                    if (state.windowCount > opts.volumeSpikeMultiplier * state.ewmaBaseline &&
                        state.ewmaBaseline > 0.0) {
                        reputation.recordRreqFloodAttempt(sourceUhid)
                    }
                }

                // Start fresh window
                state.windowStart = timestampMs
                state.windowCount = 1
            } else {
                state.windowCount += 1
            }
        }
    }

    private fun checkScatter(sourceUhid: String, destinationUhid: String, timestampMs: Long) {
        val lock    = scatterLocks.computeIfAbsent(sourceUhid) { Any() }
        val entries = scatterData.computeIfAbsent(sourceUhid) { mutableListOf() }

        synchronized(lock) {
            entries.add(Pair(destinationUhid, timestampMs))

            // Prune entries outside the rolling window
            entries.removeAll { timestampMs - it.second > opts.scatterWindowMs }

            // Count unique destinations still inside the window
            val uniqueDests = entries.map { it.first }.toSet().size
            if (uniqueDests > opts.scatterThreshold) {
                reputation.recordRreqFloodAttempt(sourceUhid)
            }
        }
    }
}
